packages/react-dom/src/__tests__/ReactDOMEventListener-test.js

```diff
@@ -119,41 +129,59 @@ describe('ReactDOMEventListener', () => {
             }
           }
         }
-        ReactDOM.render(<MyComponent />, container);
-        container.firstChild.dispatchEvent(
-          new MouseEvent('click', {
-            bubbles: true,
-          }),
-        );
+        const root = ReactDOMClient.createRoot(container);
+        await act(() => {
+          root.render(<MyComponent />);
+        });
+        await act(() => {
+          container.firstChild.dispatchEvent(
+            new MouseEvent('click', {
+              bubbles: true,
+            }),
+          );
+        });
         expect(container.firstChild.textContent).toBe('clicked!');
       } finally {
         document.body.removeChild(container);
       }
     });
 
-    it('should batch between handlers from different roots', () => {
+    it('should batch between handlers from different roots', async () => {
```

Comment:
I think this is not quite faithfully reproducing what a browser would do.

What this is testing is that we have two nested containers and that each container has event delegation listeners on them.

The event will bubble through both and so trigger two event delegation events into React. Ideally they'd be batched but we need some time to flush them.

For some events we can flush in a macro task such as at rAF. However, for discrete events this is too late. There is no event that is guaranteed to fire later than the current container event listener and before the next discrete event. Ideally on the `window` but this is not guaranteed to fire at all if propagation is stopped.

So for discrete events we use microtask. Unfortunately, IIRC, microtasks fire between every bubbled event. When you manually dispatchEvent this doesn't happen because it's already in a macrotask but the browser's dispatch schedules separate macrotasks for each bubbled event. Therefore for a discrete flush it should flush between the two different events.

You're not giving it an opportunity to flush microtasks between the bubbling which the browser does. So to simulate that you might have to manually fire the bubbling and yield with a couple of awaits between each firing. To do this correctly you might have to patch the target on the event object.

Now, for this particular test I don't know exactly the semantics because it's using a mouseout event which is not a discrete event so even with the right testing it might "work" by batching.

Therefore I think we should make a duplicate of this same test with a discrete event too like click because it should definitely not work there.

Reply:
Ok, I reverted the changes to the test in this PR, and followed up with a fix using the strategy you outlined here: https://github.com/facebook/react/pull/28079

Also added a separate test so we cover both discrete and continuous, which shows the different expected behavior.